room.go

```diff
@@ -311,7 +314,15 @@ func (r *Room) JoinWithToken(url, token string, opts ...ConnectOption) error {
 				logger.Debugw("RTC engine joining room",
 					"url", bestURL,
 				)
-				joinRes, err = r.engine.Join(bestURL, token, params)
+				// Not exposing this timeout as an option for now so that callers don't
+				// set unrealistic values.  We may reconsider in the future though.
+				// 4 seconds chosen to balance the trade-offs:
+				// - Too long, users will given up.
+				// - Too short, risk frequently timing out on a request that would have
+				//   succeeded.
+				callCtx, cancelCallCtx := context.WithTimeout(ctx, 4 * time.Second)
+				joinRes, err = r.engine.JoinContext(callCtx, bestURL, token, params)
+				cancelCallCtx()
```

Comment:
guessing with exponential backoff in room.go, it could still time out. I think it does a bunch of tries. Should we limit that also to be within like 15 seconds total maybe?